qiling/extensions/r2/r2.py

```diff
@@ -0,0 +1,162 @@
+
+
+@dataclass(unsafe_hash=True, init=False)
+class Function(R2Data):
+    name: str
+    offset: int
+    size: int
+    signature: str
+
+
+@dataclass(unsafe_hash=True, init=False)
+class Section(R2Data):
+    name: str
+    size: int
+    vsize: int
+    paddr: int
+    vaddr: int
+    perm: str  # TODO: use int or enum
+
+
+@dataclass(unsafe_hash=True, init=False)
+class String(R2Data):
+    string: str
+    vaddr: int
+    paddr: int
+    size: int
+    length: int
+    section: str = None
+
+
+@dataclass(unsafe_hash=True, init=False)
+class Symbol(R2Data):
+    # see https://github.com/rizinorg/rizin/blob/dev/librz/include/rz_bin.h
+    class SymbolType(str, Enum):
+        NOTYPE = "NOTYPE"
+        OBJ = "OBJ"
+        FUNC = "FUNC"
+        FIELD = "FIELD"
+        IFACE = "IFACE"
+        METH = "METH"
+        STATIC = "STATIC"
+        SECT = "SECT"
+        FILE = "FILE"
+        COMMON = "COMMON"
+        TLS = "TLS"
+        NUM = "NUM"
+        LOOS = "LOOS"
+        HIOS = "HIOS"
+        LOPROC = "LOPROC"
+        HIPROC = "HIPROC"
+        SPCL = "SPCL"
+        UNK = "UNK"
+
+    class SymbolBind(str, Enum):
+        LOCAL = "LOCAL"
+        GLOBAL = "GLOBAL"
+        WEAK = "WEAK"
+        NUM = "NUM"
+        LOOS = "LOOS"
+        HIOS = "HIOS"
+        LOPROC = "LOPROC"
+        HIPROC = "HIPROC"
+        IMPORT = "IMPORT"
+        UNKNOWN = "UNKNOWN"
+
+    name: str
+    realname: str
+    bind: str
+    size: int
+    type: SymbolType
+    vaddr: int
+    paddr: int
+    is_imported: bool
+
+
+class R2:
+    def __init__(self, ql: Qiling, baseaddr=(1 << 64) - 1, loadaddr=0):
+        super().__init__()
+        self.ql = ql
+        self.baseaddr = baseaddr  # r2 -B [baddr]   set base address for PIE binaries
+        self.loadaddr = loadaddr  # r2 -m [addr]    map file at given address
+
+        self._r2c = libr.r_core.r_core_new()
+        if ql.code:
+            self._setup_code()
+        else:
+            self._setup_file()
+
+        # set architecture and bits for r2 asm
+        self._cmd(f"e,asm.arch={ql.arch.type.name.lower().removesuffix('64')},asm.bits={ql.arch.bits}")
+
+    def _setup_code(self):
+        sz = len(self.ql.code)
+        path = f'malloc://{sz}'.encode()
+        fh = libr.r_core.r_core_file_open(self._r2c, path, UC_PROT_ALL, self.loadaddr)
+        libr.r_core.r_core_bin_load(self._r2c, path, self.baseaddr)
+        cmd = f'wx {self.ql.code.hex()}'
+        self._cmd(cmd)
+
+    def _setup_file(self):
+        path = self.ql.path.encode()
+        fh = libr.r_core.r_core_file_open(self._r2c, path, UC_PROT_READ | UC_PROT_EXEC, self.loadaddr)
+        libr.r_core.r_core_bin_load(self._r2c, path, self.baseaddr)
+
+    def _cmd(self, cmd: str) -> str:
+        r = libr.r_core.r_core_cmd_str(
+            self._r2c, ctypes.create_string_buffer(cmd.encode("utf-8")))
+        return ctypes.string_at(r).decode('utf-8')
+
+    def _cmdj(self, cmd: str) -> list[dict]:
```

Comment:
`list` and `dict` type annotations became available starting from Python 3.9.
We officially (?) support 3.8, so we'll have to use `typing.List` and `typing.Dict` respectively.